arngll/src/frame_info.rs

```diff
@@ -0,0 +1,690 @@
+        dst_addr: HamAddr::EMPTY,
+        src_addr: HamAddr::EMPTY,
+        rly_addr: None,
+        sec_info: None,
+        ack_crc: 0,
+    };
+
+    pub fn ack_calc(&self, payload: &[u8]) -> Option<(u16,HamAddr)> {
+        if self.ack_requested {
+            let ack_sender = if let (Some(rly_addr),true) = (self.rly_addr, self.is_from_relay) {
+                rly_addr
+            } else {
+                self.dst_addr
+            };
+            Some((self
+                    .bytes_with_payload(payload)
+                    .fold(X25.digest(), |mut digest, x| {
+                        digest.update(&[x]);
+                        digest
+                    })
+                    .finalize(),
+                  ack_sender,
+            ))
+        } else {
+            None
+        }
+    }
+
+    pub fn generate_ack_frame(&self, payload: &[u8]) -> Option<FrameInfo> {
+        if let Some((ack_crc, src_addr)) = self.ack_calc(payload) {
+            Some(FrameInfo {
+                frame_type: FrameType::Ack,
+                src_addr,
+                ack_crc,
+                .. FrameInfo::EMPTY
+            })
+        } else {
+            None
+        }
+    }
+    
+    pub fn try_from_bytes(frame: &[u8]) -> Result<(FrameInfo, &[u8]), Error> {
+        if frame.len() < 5 {
+            bail!("Frame too small");
+        }
+        let mut iter = frame.into_iter();
+
+        let fcb_msb = iter.next().copied().ok_or_else(||format_err!("Frame too small"))?;
+        let ver = fcb_msb >> 6;
+
+        if ver != VERSION_EXPERIMENTAL && ver != VERSION_1 {
+            bail!("Unexpected version {}", ver);
+        }
+
+        let dst_len = ((((fcb_msb & 0b1100) >> 2) + 1) * 2) as usize;
+        let frame_type = FrameType::try_from_u8((fcb_msb & 0b00110000) >> 4).unwrap();
+
+        let (
+            has_security_header,
+            has_netid,
+            ack_requested,
+            has_rly_addr,
+            is_from_relay,
+            rly_len,
+            has_dst_addr,
+        ) = if frame_type != FrameType::Ack {
+            let lsb = iter.next().copied().ok_or_else(||format_err!("Frame too small"))?;
+
+            (
+                (lsb & 0b10000000) != 0,
+                (lsb & 0b01000000) != 0,
+                (lsb & 0b00100000) != 0,
+                (lsb & 0b00010000) != 0,
+                (lsb & 0b00001000) != 0,
+                (((lsb & 0b11) + 1) * 2) as usize,
+                true,
+            )
+        } else {
+            (
+                false,
+                false,
+                false,
+                false,
+                false,
+                0,
+                false,
+            )
+        };
+
+        let network_id = if has_netid {
+            Some(NetworkId::from_iter(&mut iter))
+        } else {
+            None
+        };
+
+        let dst_addr = if has_dst_addr {
+            let dst_addr = HamAddr::try_from_slice(&iter.as_slice()[..dst_len])?;
```

Comment:
This pattern of treating the iterator as a slice, and then advancing it by the amount consumed, recurs in several places here. I wonder if `.take()` could be usefully employed here, instead, perhaps obviating the `.as_slice` method. E.g.,`

    let dst_addr = HamAddr::try_from_slice(&iter.take(dst_len).collect())?;

I admit I haven't tried it, and it may start to smell bad due to excessive turbofishing.